### DISCUSSION TITLE
Test cases are necessary to keep the API compatibility and overall stability.

### DISCUSSION THREAD

**Discussion by Readon** [SpinalHDL/SpinalHDL#931] [Ideas]
Recently, I always met the problems caused by API changing.
However, it's hard for reviewer to guarantee the correctness of all PR, remember all API existing is really hard. 
We might need more test cases to keep the API consistency and compatibility.

So, it's best to commit the test case while an issue is fired.
Then, if the test is merged, then the related API would not be changed in a period.

**Comment by numero-744** [comment]
API back-compatibility is one of the main goals of https://github.com/SpinalHDL/SpinalDoc-RTD/issues/123

It is true that API must not break without a period of deprecation, and a version change depending on the versioning system used.

The Rust compiler team commits the test cases from issues; maybe we should do so add a file for each issue (eg: `Issue931.scala`), based on the following template:

```scala
package spinal.tester.scalatest

import org.scalatest.funsuite.AnyFunSuite

import spinal.core._
import spinal.core.sim._

class Issue931 extends AnyFunSuite {
  test("Minimal example") {
    class Dut extends Component {
      val io = new Bundle {
        // Component interface
      }
      // Component behavior
    }

    SimConfig.withWave.doSim(new Dut) { dut =>
      // Simulation code
    }
  }
}
```

Related to #902 #619 #909

Let me play a little bit with `in`/`out`/`inout`/`master`/`slave`, then I go back to #909 :wink:

**Comment by Dolu1990** [comment]
Yes right, as much as possible, avoiding breaking others people stuff, don't want to waste others time.

**Reply by likewise** [reply]
I tried writing a unit test for issue #963. See the test there.

However, how do I run the test?

I have tried a lot, but my unit test seems not to run.

**Reply by numero-744** [reply]
Does this work?

```sh
sbt "testOnly spinal.tester.scalatest.Issue963"
```

**Comment by numero-744** [comment]
Spinal is a great library :sparkles: which has the chance to be actively maintained and contributed to :heart: so we can notice today a few API design mistakes from the past¹, and PRs merged now have a non-0 probability to trigger the need of breaking changes in the long term.

So there are cases when breaking changes are actually needed. Therefore on the long term we cannot just say "never do breaking changes". We should define, document and adopt a policy about breaking changes.

1. For instance we could adopt a versioning policy like [Semantic Versioning](https://semver.org/), which states that breaking changes are done only on major releases (when the version is `major.minor.patch`). This way users know when they update their dependency that there might be breaking changes.
2. Also, the way we integrate code today will have an impact on breaking changes on the long term. Declaring internals private, reviewing, checking against code warnings while reviewing², preventing pushes on `dev` branches without a PR, etc. are examples of things that can help reducing the need of breaking changes.

Even with a lot of attention to prevent breaking changes as much as possible, we cannot provide a guarantee to never need breaking changes ever. The thing that we must provide is a long enough deprecation period with useful help to let users time to update their code, and not just change things without giving the user clue about the issue.

Once the user has information to anticipate (versioning policy + deprecation warnings), API-breaking-changes-related errors can be really easy to perform; sometimes it is just replacing the name of something. For instance, it took me only few minutes to update *all* the SpinalHDL code to add braces after `Bool` when needed³. 

---

¹: For instance, the supposedly helper function `inWithNull` is actually a new unwanted direction type, which is a design mistake that should be fixed on the long term. For now it is deprecated (as `in` now manages `null` as well) but `inWithNull` is still reporting uncovered cases when matching over direction types, which is an issue we should fix on the long term for instance by removing `inWithNull`. For the user who has the information, it is really simple to update, just replace the function name by `in`, done!
²: It could be integrated in the GitHub workflow but there are currently warnings in the code base¹ (which is not a good thing), I had fixed the easy ones but for the remaining ones I didn't know the code enough.
³: What I did is temporarily removing the function, put `Bool()` in the clipboard, then for each error in the error list reported by my IDE, double-click to jump to the error in the code and Ctrl-V. Scroll, double-click, Ctrl-V

**Reply by Readon** [reply]
Some distribution have a mechanism that maintains several active releases, which keeps API consistency in minor version individually. 
That would need more man power to keep those.

Recall the pain that Python2 to Python3's, Scala2 to Scala3's migration, enough to rain a small community.

**Reply by numero-744** [reply]
Distributions maintaining packages != having a manifest file given to a build tool, for each project, explicitly providing a version number for each library dependency.

So I think Python and Scala are bad examples, they are not libraries.

**Reply by Dolu1990** [reply]
> which is an issue we should fix on the long term for instance by removing inWithNull

- Keeping it in break nobody's code
- Removing it will break people code, force them to dig into it, making them lose time, even if that's just replacing it with "in", things can quite easily escalate into hours lost (ex you didn't typed that code yourself, you don't know much about the project structure / version, you don't have too much control about versionning tools and so on, you need to open a PR, ...).

Time is very very valuable, for which reason should we delete inWithNull ?
Deprecating it may be enough.

**Reply by numero-744** [reply]
It is deprecated but as stated earlier it triggers warnings. It should at least be transformed into a real helper functions, to stop it from being a non-sense direction.
